### ISSUE TITLE
Convex geometry of the hyperbolic plane

### ISSUE THREAD

**Issue by sfreedman67** [flatsurf/sage-flatsurf#156]
This is a draft of notes that we took at a meeting in Bordeaux in early 2022. Please extend these notes.

---

Geometry on the hyperbolic plane is not really implemented in SageMath. It's needed for our implementation of #155. While it might eventually go upstream into SageMath, we want to first implement it in sage-flatsurf.

Here are some objects that we think we should implement:

```python
class flatsurf.geometry.HyperbolicPlane:
    # over a fixed base ring. Probably does nothing but store that base ring.

class flatsurf.geometry.HyperbolicBoundaryPoint:
    # projective coordinates of a boundary point.

class flatsurf.geometry.HyperbolicGeodesic:
    
class flatsurf.geometry.HyperbolicEdge:
    # a geodesic segment

class flatsurf.geometry.HyperbolicHalfPlane:
    # a geodesic + an orientation. Implements sagemath containment for points.
    
class flatsurf.geometry.HyperbolicPoint:
    # a pair of geodesics. lazily computes cached exact coordinates, e.g., for hashing.
    
class flatsurf.geometry.HyperbolicPolytope:
    # intersection of half planes, as a list of edges.
    
```

It is not completely clear which hyperbolic model is the best to work with (probably the hyperboloid is more "linear").

**Comment by videlec** [comment]
I also think that a boundary point is a special case of point (it is just that the quadratic form evaluates to zero). There is also "beyond ideal" points that are in the positive cone of the space.

**Comment by saraedum** [comment]
Fixed in #158.
